### PR TITLE
improve pubgrub error messages

### PR DESCRIPTION
**Pull request by tomerd**
motivation: clearer error messages help users become more self sufficient

changes:
* simplify language of error messages to be clearer
* update error messages about incompatible tools version to include the tools version in question
* adjust tests

rdar://63732571

### CODE REVIEW

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test os x

**Comment by tomerd** [comment]
@swift-ci please test linux

**Comment by tomerd** [comment]
@swift-ci please test self hosted linux

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift
Comment (abertelrud) [inline Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift:1420]: I think the `::` form a lot more than the bracket form, which never made sense to me.

File: Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift
Comment (abertelrud) [inline Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift:935]: Is the word "practically" necessary here?

File: Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift
Comment (abertelrud) [inline Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift:846]: I wonder if it would be more clear to lead with the incompatibility rather than putting it at the very end, but the new wording in the composed string is definitely an improvement, and I don't know that turning the order around would work for all cases.  But it might be something to consider for a future proposal.

In an IDE, each of these clauses should ideally be clickable where it takes you to the range of text in the package manifest that declares each dependency.  But we're far from having enough information to show that yet.

**Review by abertelrud — APPROVED** [review]

File: Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift
Comment (tomerd) [reply Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift:935]: this one is interesting. it is used in two situations afaict:
1. when there is a transitive dependency 
2. when there is a constraint (tools version, etc) that "forces" the resolver to use a specific version(s).

I originally changed this to "transitively depends on" which did not fit the 2nd case, so changed it to "practically" which I dont love but I feel is better than the original "requires" as it at least tells you its deducted. totally open to better language tho

File: Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift
Comment (tomerd) [reply Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift:846]: > In an IDE, each of these clauses should ideally be clickable where it takes you to the range of text in the package manifest that declares each dependency. But we're far from having enough information to show that yet.

totally agree, unfortunately it does not look like the infrastructure to point the source location is in place to support this 

> I wonder if it would be more clear to lead with the incompatibility rather than putting it at the very end, but the new wording in the composed string is definitely an improvement, and I don't know that turning the order around would work for all cases. But it might be something to consider for a future proposal.

Yes, I also feel the same way. I will try to see if I can tweak this further

File: Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift
Comment (abertelrud) [reply Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift:935]: I see.  Maybe "effectively", although I can't really say it's any more obvious than "practically".  We can always tweak it later, since these are just display strings and it shouldn't hold up the PR.

File: Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift
Comment (abertelrud) [reply Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift:846]: Right, the infrastructure for capturing locations is definitely not there yet.  One approach I've wanted to experiment with would be to add an implicit `#line` argument to each of the declarations in the PackageDescription module so that we would at least get line information, and then pass that back through the JSON to SwiftPM.  I haven't yet found time to try that out but I think that would give us at least rough locations in the manifest for each of the creations of the dependencies.  This could also be used for creation of targets, products, etc, and other objects in the manifest.  It might also be possible to build the manifest for debugging and then get callsites that way, but it doesn't seem trivial either way.

**Comment by abertelrud** [comment]
This seems good to merge, unless you want Boris' review as well.

File: Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift
Comment (tomerd) [reply Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift]: > I wonder if it would be more clear to lead with the incompatibility rather than putting it at the very end, but the new wording in the composed string is definitely an improvement, and I don't know that turning the order around would work for all cases. But it might be something to consider for a future proposal.

@abertelrud I took another (and more aggressive) swing at it, now reversing the order of how the "argument" is built. changes to test cases verification text probably are best indication of the outcome

**Comment by tomerd** [comment]
> This seems good to merge, unless you want Boris' review as well.

yes lets wait for @neonichu feedback as well since this may be sensitive

File: Tests/PackageGraphTests/PubgrubTests.swift
Comment (abertelrud) [inline Tests/PackageGraphTests/PubgrubTests.swift:1297]: Was this left by mistake?

File: Tests/PackageGraphTests/PubgrubTests.swift
Comment (abertelrud) [inline Tests/PackageGraphTests/PubgrubTests.swift:1305]: Does this indicate a footnote to the second line of the error message?  If so, I'll think about a way in which this could be annotated in the DiagnosticsEngine so that an IDE could display it indented under that line.

File: Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift
Comment (abertelrud) [reply Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift]: That's great — I think this is a big improvement.  The messages are still pretty gnarly in many cases, but I think it looks a lot better.

File: Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift
Comment (abertelrud) [inline Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift:1416]: I like the simplification of not showing the product name when it is the same as the package name, but could it lead to confusion if some products are shown and others not?  It will not be possible to distinguish the "no-product" case visually from the case where the product has the same name as the package.

Perhaps this isn't actually a problem in practice, and the simplification certainly makes things read better.  I think that the product name started being listed there as part of the changes for target-based dependencies.

Comment (tomerd) [reply Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift:1416]: good point @abertelrud. been thinking about this more and I now suspect we can drop the product name in all cases, including when the package and product names are different (which is the common case). since the error message explains why specific combination of package versions can't work together, I am not convinced the product names are material to understanding the issue - its more of additional/supplemental information resulting from target based deps - and as such I dont think worth the clutter. I proposed we remove it in all cases.

File: Tests/PackageGraphTests/PubgrubTests.swift
Comment (tomerd) [reply Tests/PackageGraphTests/PubgrubTests.swift:1305]: yes

File: Tests/PackageGraphTests/PubgrubTests.swift
Comment (tomerd) [reply Tests/PackageGraphTests/PubgrubTests.swift:1297]: yes :D

File: Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift
Comment (tomerd) [reply Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift]: >  The messages are still pretty gnarly in many cases

I agree, but that would require much deeper refactoring of the resolver engine

File: Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift
Comment (abertelrud) [reply Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift]: I agree.  I'm not sure of the reasoning for why it was added, but since the dependencies are between packages, I agree it can be removed.

**Review by abertelrud — APPROVED** [review]

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift
Comment (abertelrud) [reply Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift]: Absolutely, that's for another PR.

**Review by neonichu — APPROVED** [review]

**Comment by abertelrud** [comment]
Since @neonichu has approved, I'm merging this.

**Pull request merged**
